File: src/implement/GameDevelop.java
package implement;

import java.io.*;

public class GameDevelop {
    static int direction = 0;
    public static void main(String[] args) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] input = br.readLine().split(" ");
        int N = Integer.parseInt(input[0]), M = Integer.parseInt(input[1]);

        String[] input2 = br.readLine().split(" ");
        int x = Integer.parseInt(input2[0]), y = Integer.parseInt(input2[1]);
        direction = Integer.parseInt(input2[2]);

        int[][] arr = new int[N][M], visited = new int[N][M];

        for (int i = 0; i < N; i++) {
            String[] nums = br.readLine().split(" ");
            for (int j = 0; j < M; j++) {
                arr[i][j] = Integer.parseInt(nums[j]);
                visited[i][j] = 0;
            }
        }

        visited[x][y] = 1;

        int[] dx = {-1, 0, 1,0}, dy = {0, 1, 0, -1};

        int turnTime = 0, count = 1;

        while(true){
            turnLeft();

            int nx = x + dx[direction], ny = y + dy[direction];

            if(visited[nx][ny] == 0 && arr[nx][ny] == 0){
                x = nx;
                y = ny;
                count++;
                turnTime = 0;
                continue;
            }else{
                turnTime++;
            }

            if(turnTime == 4){
                x = nx;
                y = ny;
            }else{
                break;
            }

            turnTime = 0;
        }

        bw.write(String.valueOf(count));
        bw.flush();
        bw.close();
    }

    private static void turnLeft(){
        direction -= 1;

        if(direction == -1) direction = 3;
    }
}

File: JS/implement/gameDevelop.js
"use strict";

const fs = require("fs");

const input =
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString().split("\n")
    : `4 4
1 1 0
1 1 1 1
1 0 0 1
1 1 0 1
1 1 1 1`.split("\n");

const [N, M] = input[0].split(" ").map((v) => parseInt(v));

const arr = Array.from(Array(N), () => new Array(M));
const visited = Array.from(Array(N), () => new Array(M).fill(0));
let count = 0;

let [x, y, direction] = input[1].split(" ").map((v) => parseInt(v));

visited[x][y] = 1;

for (let i = 2; i < N + 2; i++) {
  arr[i - 2] = input[i].split(" ").map((v) => parseInt(v));
}

const dx = [-1, 0, 1, 0];
const dy = [0, 1, 0, -1];

let turnTime = 0;

function turnLeft() {
  direction -= 1;
  if (direction === -1) direction = 3;
}

while (true) {
  turnLeft();

  let nx = x + dx[direction];
  let ny = y + dy[direction];

  if (visited[nx][ny] === 0 && arr[nx][ny] === 0) {
    x = nx;
    y = ny;
    count++;
    turnTime = 0;
    continue;
  } else {
    turnTime++;
  }

  if (turnTime === 4) {
    nx = x - dx[direction];
    ny = y - dy[direction];

    if (arr[nx][ny] === 0) {
      x = nx;
      y = ny;
    } else {
      break;
    }
    turnTime = 0;
  }
}

console.log(count);
